consensus/obcpbft/pbft-core.go

```diff
@@ -249,67 +256,60 @@ func newPbftCore(id uint64, config *viper.Viper, consumer innerStack) *pbftCore
 
-		select {
-		case <-instance.closed:
-			close(instance.idleChan)
-			return
-		case <-instance.newViewTimer.C:
-			logger.Info("Replica %d view change timer expired, sending view change", instance.id)
-			instance.sendViewChange()
-		case msg := <-instance.incomingChan:
-			logger.Debug("Replica %d received incoming message from %v", instance.id, msg.sender)
-			instance.recvMsg(msg.msg, msg.sender)
-		case update := <-instance.stateUpdatingChan:
-			instance.skipInProgress = true
-			instance.lastExec = update.seqNo
-			instance.moveWatermarks(instance.lastExec) // The watermark movement handles moving this to a checkpoint boundary
-		case update := <-instance.stateUpdatedChan:
-			seqNo := update.seqNo
-			logger.Info("Replica %d application caught up via state transfer, lastExec now %d", instance.id, seqNo)
-			// XXX create checkpoint
-			instance.lastExec = seqNo
-			instance.moveWatermarks(instance.lastExec) // The watermark movement handles moving this to a checkpoint boundary
-			instance.skipInProgress = false
-			instance.executeOutstanding()
-		case <-instance.execCompleteChan:
-			instance.execDoneSync()
-		case work := <-instance.injectChan:
-			work() // Used to allow the caller to steal use of the main thread
-		case instance.idleChan <- struct{}{}:
-			// Used to detect when this thread is idle for testing
-		}
+	switch e.eventType() {
+	case viewChangeTimerEventID:
+		logger.Info("Replica %d view change timer expired, sending view change", instance.id)
+		instance.timerActive = false
+		instance.sendViewChange()
+	case messageEventID:
+		msg := e.(messageEvent)
+		logger.Debug("Replica %d received incoming message from %v", instance.id, msg.sender)
+		instance.recvMsg(msg.msg, msg.sender)
+	case stateUpdatingEventID:
+		update := e.(stateUpdatingEvent)
+		instance.skipInProgress = true
+		instance.lastExec = update.seqNo
+		instance.moveWatermarks(instance.lastExec) // The watermark movement handles moving this to a checkpoint boundary
+	case stateUpdatedEventID:
+		update := e.(stateUpdatedEvent)
+		seqNo := update.seqNo
+		logger.Info("Replica %d application caught up via state transfer, lastExec now %d", instance.id, seqNo)
+		// XXX create checkpoint
+		instance.lastExec = seqNo
+		instance.moveWatermarks(instance.lastExec) // The watermark movement handles moving this to a checkpoint boundary
+		instance.skipInProgress = false
+		instance.executeOutstanding()
+	case execDoneEventID:
+		instance.execDoneSync()
+	case workEventID:
+		e.(workEvent)() // Used to allow the caller to steal use of the main thread, to be removed
+	default:
+		logger.Error("Replica %d received an unknown message type", instance.id)
 	}
+
+	return nil
 }
 
 // Allows the caller to inject work onto the main thread
 // This is useful when the caller wants to safely manipulate PBFT state
 func (instance *pbftCore) inject(work func()) {
-	instance.injectChan <- work
-	<-instance.idleChan
+	instance.manager.queue() <- workEvent(work)
```

Comment:
I am looking for some idiomatic go, not what our (sometimes a bit java-y) codebase does :)